### PR TITLE
feat(lsp): defaults: tagfunc, omnifunc

### PR DESCRIPTION
**Pull request by mfussenegger**
Opening this up as draft to test the waters.

The main motivation is to make the LSP setup a bit easier. Currently `:help lsp-config` recommends users to setup `tagfunc`, `omnifunc`, and several keymaps. This is boilerplate that we may be able to avoid if we can come up with a choice that is uncontroversial enough. I think it would be acceptable if it removes boilerplate for ~90% of the use-cases, while maybe adding a bit of extra config for the remaining 10%.

A few of those are fairly unintrusive. That makes them candidates to start with. This PR starts with two:

- `omnifunc`: Is set if empty and if the server has the capability. I don't see any downsides to this.
- `tagfunc`: Is set if empty and if the server has the capability. This overrides regular ctags uses so that LSP symbols have priority. ctags will still be used as fallback if the LSP server times out. If a user wants to continue using regular ctags without querying the language server, they'd have to run `setlocal tagfunc<` in a `LspAttach` event.

The options are set shortly before `LspAttach` & `config.on_attach` gets triggered. This should ensure that custom configuration in `FileType` events would already have happened (so those are *not* overridden), and also give the option to *undo* the settings.

### Other options:

- Leave things as they are
- Add a `vim.lsp.set_defaults` function or something like that. Users could call that within a `LspAttach` event or the `on_attach` handler. (I fear this might explode into some kind of custom keymap DSL if users start requesting to have a subset of defaults) (*edit*: We'd probably need a `unset_defaults` to use for detach as well?)
- Add this, but only with explicit opt-in. E.g. add a `set_default` flags to `vim.lsp.start`.
- Add this, but with an opt-out. E.g. add a `set_default` flags to `vim.lsp.start`.

### Possible follow up:

- keywordprg or map K to hover
- formatexpr

This draft intentionally left out docs updates and detach logic. I'll add that only if this has a chance of moving forward.

### CODE REVIEW

**Comment by clason** [comment]
Personal opinion: unalloyed 👍 for the two defaults here.

Regarding other options:

* strong preference for opt-in over opt-out
* _very_ strong preference for "take it or leave it"; no further configurability than opting into an opinionated set of "quick start settings"
* adding a `set_defaults` (default `false`) to `vim.lsp.start` is the simplest option, but a `vim.lsp.set_defaults` mapping that can just be passed as `on_attach = vim.lsp.set_defaults` is more flexible, since it can also be used 
   - with `nvim-lspconfig`
   - inside your own `on_attach` function
   - in an `LspAttach` autocommand

File: runtime/lua/vim/lsp.lua [hunk 1]
```diff
@@ -1299,6 +1299,13 @@ function lsp.start_client(config)
   function client._on_attach(bufnr)
     text_document_did_open_handler(bufnr, client)
 
+    if client.server_capabilities.definitionProvider and vim.bo[bufnr].tagfunc == '' then
+      vim.bo[bufnr].tagfunc = 'v:lua.vim.lsp.tagfunc'
+    end
+    if client.server_capabilities.completionProvider and vim.bo[bufnr].omnifunc == '' then
+      vim.bo[bufnr].omnifunc = 'v:lua.vim.lsp.omnifunc'
+    end
```
Comment (justinmk) [inline runtime/lua/vim/lsp.lua:1307]: should we start a set_defaults (private) function in case we continue adding things in the future?

**Review by justinmk — APPROVED** [review]
Let's do it. Need a mention at `:help lsp-config`

**Review by mfussenegger — COMMENTED** [review]
Updated the docs in lsp-config and extracted the logic into set_defaults and added a unset_defaults.

This should now be good to go.

**Comment by clason** [comment]
Let's give it a whirl then!

**Pull request merged**